### PR TITLE
UR20 description and meshes

### PR DESCRIPTION
**Pull request by urrsk**
The UR20 meshes are added under Universal Robots A/S’ Terms and Conditions for Use of Graphical Documentation

### CODE REVIEW

**Comment by fmessmer** [comment]
thanks for the URDF and meshes :tada: 
they work great for us....could we get this merged - and released to Noetic - some time soon?

File: ur20_moveit_config/CMakeLists.txt
Comment (fmauch) [inline ur20_moveit_config/CMakeLists.txt]:
```suggestion
cmake_minimum_required(VERSION 3.0.2)
```

I know, the UR16e also still has 2.8.3, this should be fixed, as well. But not as part of this PR, obviously.

File: ur_description/package.xml
Comment (fmauch) [inline ur_description/package.xml:3]:
```suggestion
<package format="2">
```

I think for now it makes sense to stick with package format 2 until https://github.com/ros-infrastructure/rep/pull/347 is resolved.

**Review by fmauch — COMMENTED** [review]
Maybe we also want to add a README explaining the licensing situation with the UR20 meshes as we did it for the ROS 2 description?

@gavanderhoorn I feel like this would be a sensible point to create a `noetic-devel` branch. Technically, this would probably be possible to build and use on ROS noetic, as well, but since melodic isn't actively maintained anymore I feel like adding new features to a `melodic-devel` branch seems wrong.

File: ur_description/package.xml
Comment (urrsk) [reply ur_description/package.xml:3]: As far as I remember, the reason why we change the format version was to support multiple licenses

Comment (fmauch) [reply ur_description/package.xml:3]: The full license to files support isn't there, yet, hence it is in the commented line. Having multiple license tags is also possible in format 2.

Comment (urmahp) [reply ur_description/package.xml:3]: As far as I remember then the reason for changing the package format, is to be able to use the following in the license.
```
<license file="meshes/ur20/LICENSE.txt">Universal Robots A/S' Terms and Conditions for Use of Graphical Documentation</license>
```
 The license file is not supported in package format 2 only in 3.

Comment (fmauch) [reply ur_description/package.xml:3]: I remembered that exactly that was supported by format 2, as well. I checked the specs and it seems I remembered that incorrectly.

**Review by fmauch — APPROVED** [review]
With the cmake version increase of the UR16e being the only thing not necessarily belonging to this PR I woud be fine merging this. We'll just have to decide on the target branch.

**Comment by RobertWilbrandt** [comment]
I agree with merging this to `noetic-devel`. As there are no further melodic binary releases, people have to build from sources anyway to get this support and using noetic sets the right expectations about maintenance and robustness.

**Comment by fmauch** [comment]
Alright, I'll merge this on Friday then and start releasing from a `noetic` branch.

File: ur20_moveit_config/CHANGELOG.rst
Comment (gavanderhoorn) [inline ur20_moveit_config/CHANGELOG.rst:1]: Please remove. This file will/should be auto-generated by the release tooling.

File: ur20_moveit_config/launch/planning_context.launch
Comment (gavanderhoorn) [inline ur20_moveit_config/launch/planning_context.launch:16]: I'm not sure why we don't do that elsewhere, but shouldn't this load the `joint_limits.yaml` from `ur_description` instead? That was the idea at least.

File: ur_description/meshes/ur20/LICENSE.txt
Comment (gavanderhoorn) [inline ur_description/meshes/ur20/LICENSE.txt:1]: I've made my position on this license clear in private conversation.

But apparently it's necessary, so we'll leave it to individual users to decide whether they agree with it or not.

I do believe we should make the fact there is a new license clear in the readme of both the main repository *and* the readme of this package, otherwise it's too easy to miss.

The new badge is insufficient. We should add a section right at the top of the readme file drawing attention to this non-OSS license.

File: ur_description/package.xml
Comment (gavanderhoorn) [inline ur_description/package.xml:21]: please add a comment noting this license only applies to the files in `meshes/ur20`.

Just to clarify the current situation.

File: ur_kinematics/CMakeLists.txt
Comment (gavanderhoorn) [inline ur_kinematics/CMakeLists.txt]:
```suggestion
  LIBRARIES ur3_kin ur5_kin ur10_kin ur3_moveit_plugin ur5_moveit_plugin ur10_moveit_plugin ur3e_kin ur5e_kin ur10e_kin ur16e_kin ur20_kin ur20_moveit_plugin
```

the other additions are off-topic here in this PR.

File: ur_kinematics/ur_moveit_plugins.xml
Comment (gavanderhoorn) [inline ur_kinematics/ur_moveit_plugins.xml]:
```suggestion
      Original plugin developed by Kelsey Hawkins from Georgia Tech.
```

let's be specific.

File: ur_description/package.xml
Comment (gavanderhoorn) [inline ur_description/package.xml]:
```suggestion
      URDF description for Universal UR3(e), UR5(e), UR10(e), UR16e and UR20 robots
```

let's be specific about things. We list the exact models elsewhere as well.

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
First round of comments.

File: ur20_moveit_config/launch/planning_context.launch
Comment (fmauch) [reply ur20_moveit_config/launch/planning_context.launch:16]: I was having the same though, but since we do the same thing in the other packages I dropped those thoughts. Personally, I would be fine with postponing this to a follow-up PR where we change that in all MoveIt support packages.

File: README.md
Comment (fmauch) [inline README.md]: I think it makes sense to separate the license badge and the statement the following way:
```suggestion
[![License](https://img.shields.io/badge/License-Universal%20Robots%20A/S%E2%80%99%20Terms%20and%20Conditions%20for%20Use%20of%20Graphical%20Documentation-blue.svg)](https://www.universal-robots.com/legal/terms-and-conditions/terms_and_conditions_for_use_of_graphical_documentation.txt)\

[![support level: community](https://img.shields.io/badge/support%20level-community-lightgray.png)](http://rosindustrial.org/news/2016/10/7/better-supporting-a-growing-ros-industrial-software-platform)

[![support level: community](https://img.shields.io/badge/support%20level-community-lightgray.png)](http://rosindustrial.org/news/2016/10/7/better-supporting-a-growing-ros-industrial-software-platform)

[ROS-Industrial](https://wiki.ros.org/Industrial) Universal Robot meta-package. See the [ROS wiki](https://wiki.ros.org/universal_robots) page for compatibility information and other more information.

__License__
The [UR20 meshes](ur_description/meshes/ur20) constitutes “Graphical Documentation” the use of which is subject to and governed by our “[Terms and Conditions for use of Graphical Documentation](https://www.universal-robots.com/legal/terms-and-conditions/terms_and_conditions_for_use_of_graphical_documentation.txt)”.\
If you have any questions regarding the license or the license doesn't fit you use-case, please contact [legal@universal-robots.com](mailto:legal@universal-robots.com).

All other content is licensed under the BSD-3-Clause license or Apache-2.0 respectively.

```

Comment (urrsk) [reply README.md]: Good suggestion. Lets do that

**Comment by urrsk** [comment]
Please squash the commits when it is ready for getting merged.

File: README.md
Comment (fmauch) [inline README.md:15]: This is missing an empty line. Sorry, I didn't catch this in my suggestion.
```suggestion
__License__

```

image

File: ur20_moveit_config/launch/planning_context.launch
Comment (fmauch) [reply ur20_moveit_config/launch/planning_context.launch:16]: See #668

File: README.md
Comment (fmauch) [inline README.md]:
```suggestion
[![License](https://img.shields.io/badge/License-Universal%20Robots%20A/S%E2%80%99%20Terms%20and%20Conditions%20for%20Use%20of%20Graphical%20Documentation-blue.svg)](https://www.universal-robots.com/legal/terms-and-conditions/terms_and_conditions_for_use_of_graphical_documentation.txt)
```

**Comment by fmauch** [comment]
@gavanderhoorn Could you give this another review please?

**Comment by fmessmer** [comment]
> Alright, I'll merge this on Friday then and start releasing from a `noetic` branch.

what's blocking this PR from merging?

**Comment by fmauch** [comment]
> what's blocking this PR from merging?

A review requesting changes (or a re-review to be more precise, since these comments have been addressed).

**Comment by urrsk** [comment]
@gavanderhoorn it would valuable if you could fine the time in the near future to review this PR again.
I believe requested changes have been addressed. Otherwise please let me know.

File: README.md
Comment (gavanderhoorn) [inline README.md]: was this duplication intentional?

```suggestion
```

File: README.md
Comment (gavanderhoorn) [inline README.md]:
```suggestion
Note: the mesh files used by the [UR20](ur_description/meshes/ur20) model in the `ur_description` package come with a non-OSI approved, non-open-source license.
They are considered *Graphical Documentation* by Universal Robots and their use is subject to and governed by the "[Terms and Conditions for use of Graphical Documentation](https://www.universal-robots.com/legal/terms-and-conditions/terms_and_conditions_for_use_of_graphical_documentation.txt)".

Users of these packages are advised to make sure their use complies with these additional terms and conditions.
Please read the license carefully before making use of the UR20 variant in the `ur_description` package.
```

Edit: I would actually suggest to make the link here *also* point to `ur_description/meshes/ur20/LICENSE.txt`, just as the badge (see earlier comment).

I don't know what the link would become though, so I don't have a suggestion to easily accept and apply I'm afraid.

File: README.md
Comment (gavanderhoorn) [inline README.md]:
```suggestion
If you have any questions regarding this license or if this license doesn't fit your use-case, please contact [legal@universal-robots.com](mailto:legal@universal-robots.com).
```

File: ur20_moveit_config/config/joint_limits.yaml
Comment (gavanderhoorn) [inline ur20_moveit_config/config/joint_limits.yaml:1]: I'm not sure whether we've discussed this before, but should this be duplicating the joint limits from `ur_description` here (and only partially as well)?

Would support custom limits for just this MoveIt config, but not sure whether we feel that's a strong enough reason.

File: ur_description/meshes/ur20/LICENSE.txt
Comment (gavanderhoorn) [inline ur_description/meshes/ur20/LICENSE.txt:1]: :+1: to storing the license file in the repository. It's essentially a contract, and clients should have a copy of that contract if they are going to be bound by it.

File: README.md
Comment (gavanderhoorn) [inline README.md:9]: Should we instead make this point to the `ur_description/meshes/ur20/LICENSE.txt` file? The online version could be changed at any time. The files in this repository would fall under the version of the license as it is right now, in this PR. It would make sense to me to make this badge link to the version the files would fall under.

**Review by gavanderhoorn — COMMENTED** [review]
apologies for my delayed reply.

I've left a couple in-line comments, which I expect are easily resolved.

If desired, I can review again tomorrow, or early next week.

**Review by gavanderhoorn — APPROVED** [review]

**Comment by gavanderhoorn** [comment]
I've approved the changes to the previous review so as to not block this.

I've left new comments as *comments*, so they don't block.

File: README.md
Comment (fmauch) [reply README.md:9]: From what I understood, the license is structured such that the latest version always is relevant (see point 1.4)

> ... If you download a software package that includes the Graphical Documentation, your use of the Graphical Documentation will be subject to and governed by the latest version of the T&Cs, which can be found here: ...

Given that clause I think it does make sense to link to the online version.

File: ur20_moveit_config/config/joint_limits.yaml
Comment (fmauch) [reply ur20_moveit_config/config/joint_limits.yaml:1]: Yes, you are right, it doesn't make much sense to have them different atm. We've discussed earlier in this PR that it would make sense to actually re-use the limits from the description, hence #668 which currently waits for this PR to be finished first. However, since we decided to include the UR20 in #668 once this is merged, it would make sense to update the file here, yes.

File: README.md
Comment (fmauch) [reply README.md]: @urrsk as far as I remember this section is an outcome of UR's law department, so I guess I cannot say much to that.

File: README.md
Comment (gavanderhoorn) [reply README.md:9]: I'm not sure that's legal. There would be no change notification to current users, UR would be able to silently, unilaterally change the complete license without anyone being aware of it, and with all licenses I've worked with (and in the jurisdictions I've had to work with licenses), once you have a copy of the artefacts under that license, that copy and/or version will always fall under the version of the license as it was when you agreed to it.

I have seen licenses where there are provisions for updates made by the licensor, but those always include some sort of notification period, allowing current licensees to terminate their use of whatever is under the existing license if they feel they can't, or don't want to, agree to the new version.

Additionally, there is already a copy of the current version in the repository -- or at least there will be, after merging this PR.

It's best practice to include a copy of the license in all cases, even with OSS licenses. We don't do that currently in this repository, but that's an omission and should be fixed (in a later PR).

File: README.md
Comment (gavanderhoorn) [reply README.md]: The original section contained style violations and typos. Additionally, we can't have it state "our [license]", as this is a repository on the `ros-industrial` organisation. It's not owned by UR, and we are not UR.

The rest of the changes merely draw attention to the fact that legally, users should *not* use any of the files (and perhaps not even download them) under the new license until after they've determined they can and will agree to its terms.

I'm sure lawyers will be able to find fault with just about anything, but I can't really see what's wrong here. It's definitely a non-OSI approved, non-OSS license. That's just a fact, and I feel important to call out.

File: README.md
Comment (gavanderhoorn) [reply README.md:9]: The same section 1.4 also states:
> However, you may only make the Graphical Documentation or any whole or partial copies thereof available to the public via download if it is part of a downloadable software package that includes a copy of the T&Cs.

which is what this PR will result in: a copy will be included.

So it doesn't seem strange to me to then point to that copy, and not the online version.

That seems like it would just lead to confusing situations.

File: README.md
Comment (gavanderhoorn) [reply README.md]: @gavanderhoorn wrote:
> The rest of the changes merely draw attention to the fact that legally, users should _not_ use any of the files (and perhaps not even download them) under the new license until after they've determined they can and will agree to its terms.

which is exactly what the new license also states/asks (from the preamble of `v1.01`):
> The T&Cs constitute a binding agreement between you and Universal Robots. By accessing, storing, copying, sharing, opening, or otherwise using or disposing of the Graphical Documentation, you acknowledge that you are bound by the T&Cs.

Comment (urrsk) [reply README.md]: @gavanderhoorn thanks for your feedback.
Agree, the wording is made for a repo UR's Github organisation.
I will come back with another suggestion.
And yes it is not a OSI approved license

File: README.md
Comment (urrsk) [reply README.md:9]: You have a point. Though the motivation was to remove the burden from the repo owner to ensure that the license is up to date.

File: README.md
Comment (urrsk) [reply README.md]: Here is my suggestion.
It is important to let people understand that it is a proprietary license, though I also would like to highlight that it pretty open as well.
```suggestion
The [UR20 meshes](ur_description/meshes/ur20) constitute “Graphical Documentation” from Universal Robots which is subject to and governed by Universal Robots' "[Terms and Conditions for use of Graphical Documentation](https://www.universal-robots.com/legal/terms-and-conditions/terms_and_conditions_for_use_of_graphical_documentation.txt)".

Universal Robots' [Terms and Conditions for use of Graphical Documentation](https://www.universal-robots.com/legal/terms-and-conditions/terms_and_conditions_for_use_of_graphical_documentation.txt) do not fully comply with [OSI's definition of Open Source](https://opensource.org/osd/), but they do allow you to use, modify and share “Graphical Documentation”, including [UR20 meshes](ur_description/meshes/ur20), subject to certain restrictions.\
```

**Comment by fmessmer** [comment]
still eagerly waiting for this to get merged and released :roll_eyes:

**Comment by fmauch** [comment]
As discussed and requested by UR, I'll leave the license badge point to the online version and merge things as they currently are.

**Pull request merged**
